// @grapecity/spread-sheets-angular/esm2020/src/worksheet.component.mjs
/*import_begin*/
import { Component, ContentChildren, Input, } from '@angular/core';
/*import_end*/
import { ColumnComponent } from './column.component';
import * as GC from '@grapecity/spread-sheets';
import * as i0 from "@angular/core";
/*code_begin*/
export class WorksheetComponent {
    constructor() {
        this.sheet = new GC.Spread.Sheets.Worksheet('');
    }
    onAttached() {
        const sheet = this.sheet;
        const columns = this.columns;
        sheet.suspendPaint();
        sheet.suspendEvent();
        if (this.dataSource) {
            sheet.setDataSource(this.dataSource);
            columns.forEach((columnComponent, index) => {
                if (columnComponent.dataField) {
                    sheet.bindColumn(index, {
                        name: columnComponent.dataField,
                        displayName: columnComponent.headerText,
                    });
                }
            });
        }
        if (columns.length > 0) {
            sheet.setColumnCount(columns.length);
            columns.forEach((columnComponent, index) => {
                columnComponent.onAttached(this.sheet, index);
            });
        }
        sheet.resumeEvent();
        sheet.resumePaint();
    }
    getSheet() {
        return this.sheet;
    }
    ngOnChanges(changes) {
        const sheet = this.sheet;
        sheet.suspendPaint();
        sheet.suspendEvent();
        for (const changeName in changes) {
            const newValue = changes[changeName].currentValue;
            if (newValue === null || newValue === void 0) {
                continue;
            }
            switch (changeName) {
                case 'rowCount':
                    sheet.setRowCount(newValue);
                    break;
                case 'colCount':
                    sheet.setColumnCount(newValue);
                    break;
                case 'name':
                    sheet.name(newValue);
                    break;
                case 'frozenColumnCount':
                    sheet.frozenColumnCount(newValue);
                    break;
                case 'frozenRowCount':
                    sheet.frozenRowCount(newValue);
                    break;
                case 'frozenTrailingRowCount':
                    sheet.frozenTrailingRowCount(newValue);
                    break;
                case 'frozenTrailingColumnCount':
                    sheet.frozenTrailingColumnCount(newValue);
                    break;
                case 'selectionPolicy':
                    sheet.selectionPolicy(newValue);
                    break;
                case 'selectionUnit':
                    sheet.selectionUnit(newValue);
                    break;
                case 'zoom':
                    sheet.zoom(newValue);
                    break;
                case 'currentTheme':
                    sheet.currentTheme(newValue);
                    break;
                case 'defaultStyle':
                    sheet.setDefaultStyle(newValue);
                    break;
                case 'rowOutlineInfo':
                    newValue.forEach((item) => {
                        sheet.rowOutlines.group(item.index, item.count);
                    });
                    sheet.repaint();
                    break;
                case 'columnOutlineInfo':
                    newValue.forEach((item) => {
                        sheet.columnOutlines.group(item.index, item.count);
                    });
                    sheet.repaint();
                    break;
                case 'showRowOutline':
                    sheet.showRowOutline(newValue);
                    break;
                case 'showColumnOutline':
                    sheet.showColumnOutline(newValue);
                    break;
                case 'dataSource':
                    sheet.setDataSource(newValue);
                    break;
                case 'autoGenerateColumns':
                    sheet[changeName] = newValue;
                default:
                    sheet.options[changeName] = newValue;
            }
        }
        sheet.resumeEvent();
        sheet.resumePaint();
    }
    ngAfterViewInit() {
        this.columns.changes.subscribe(() => { this.onAttached(); });
    }
    ngOnDestroy() {
        const sheet = this.sheet;
        const spread = sheet ? sheet.getParent() : null;
        if (spread) {
            const sheetIndex = spread.getSheetIndex(sheet.name());
            if (sheetIndex !== void 0) {
                spread.removeSheet(sheetIndex);
            }
        }
    }
}
WorksheetComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.4.0", ngImport: i0, type: WorksheetComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
WorksheetComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.4.0", type: WorksheetComponent, selector: "gc-worksheet", inputs: { rowCount: "rowCount", colCount: "colCount", dataSource: "dataSource", name: "name", frozenColumnCount: "frozenColumnCount", frozenRowCount: "frozenRowCount", frozenTrailingRowCount: "frozenTrailingRowCount", frozenTrailingColumnCount: "frozenTrailingColumnCount", allowCellOverflow: "allowCellOverflow", frozenlineColor: "frozenlineColor", sheetTabColor: "sheetTabColor", selectionPolicy: "selectionPolicy", selectionUnit: "selectionUnit", zoom: "zoom", currentTheme: "currentTheme", clipBoardOptions: "clipBoardOptions", rowHeaderVisible: "rowHeaderVisible", colHeaderVisible: "colHeaderVisible", rowHeaderAutoText: "rowHeaderAutoText", colHeaderAutoText: "colHeaderAutoText", rowHeaderAutoTextIndex: "rowHeaderAutoTextIndex", colHeaderAutoTextIndex: "colHeaderAutoTextIndex", isProtected: "isProtected", showRowOutline: "showRowOutline", showColumnOutline: "showColumnOutline", selectionBackColor: "selectionBackColor", selectionBorderColor: "selectionBorderColor", defaultStyle: "defaultStyle", rowOutlineInfo: "rowOutlineInfo", columnOutlineInfo: "columnOutlineInfo", autoGenerateColumns: "autoGenerateColumns" }, queries: [{ propertyName: "columns", predicate: ColumnComponent }], usesOnChanges: true, ngImport: i0, template: `
        <ng-content></ng-content>
    `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.4.0", ngImport: i0, type: WorksheetComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'gc-worksheet',
                    template: `
        <ng-content></ng-content>
    `,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { columns: [{
                type: ContentChildren,
                args: [ColumnComponent]
            }], rowCount: [{
                type: Input
            }], colCount: [{
                type: Input
            }], dataSource: [{
                type: Input
            }], name: [{
                type: Input
            }], frozenColumnCount: [{
                type: Input
            }], frozenRowCount: [{
                type: Input
            }], frozenTrailingRowCount: [{
                type: Input
            }], frozenTrailingColumnCount: [{
                type: Input
            }], allowCellOverflow: [{
                type: Input
            }], frozenlineColor: [{
                type: Input
            }], sheetTabColor: [{
                type: Input
            }], selectionPolicy: [{
                type: Input
            }], selectionUnit: [{
                type: Input
            }], zoom: [{
                type: Input
            }], currentTheme: [{
                type: Input
            }], clipBoardOptions: [{
                type: Input
            }], rowHeaderVisible: [{
                type: Input
            }], colHeaderVisible: [{
                type: Input
            }], rowHeaderAutoText: [{
                type: Input
            }], colHeaderAutoText: [{
                type: Input
            }], rowHeaderAutoTextIndex: [{
                type: Input
            }], colHeaderAutoTextIndex: [{
                type: Input
            }], isProtected: [{
                type: Input
            }], showRowOutline: [{
                type: Input
            }], showColumnOutline: [{
                type: Input
            }], selectionBackColor: [{
                type: Input
            }], selectionBorderColor: [{
                type: Input
            }], defaultStyle: [{
                type: Input
            }], rowOutlineInfo: [{
                type: Input
            }], columnOutlineInfo: [{
                type: Input
            }], autoGenerateColumns: [{
                type: Input
            }] } });
